### DISCUSSION TITLE
Building a zero-diagonal adjacency matrix sample-wise from bipartite node's `edge_index`

### DISCUSSION THREAD

**Discussion by ndrwnaguib** [pyg-team/pytorch_geometric#5805] [Q&A]
Hello,

I have a problem concatenating`edge_index` and `edge_attr` across batches, I have a custom impl that works only after I have loaded the batch. Here is the function I use:

    def _pad_adj(self, edge_index, edge_attr, num_left_feat, num_right_feat, fill=-9e15):
        row, col = edge_index
        edge_attr = edge_attr.reshape(-1)
        row_features = sp.coo_matrix(
            (edge_attr, (row, col)),
            (num_right_feat, num_left_feat),
        )

        Axb = row_features.todense().T
        num_nodes = num_left_feat + num_right_feat
        adj = fill * np.ones((num_nodes, num_nodes))
        adj[:num_vars, num_vars:] = Axb
        adj[num_vars:, :num_vars] = Axb.T

        assert (adj[:num_left_feat, :num_left_feat] == fill).all(), "adjacency is ill-constructed"
        assert (adj[num_left_feat:, num_left_feat:] == fill).all(), "adjacency is ill-constructed"

        adj = torch.FloatTensor(adj)
        return adj.to_sparse_coo()

The above implementation produces the intended results only if `edge_index` and `edge_attr` were passed after the batch was loaded. But, this is too slow to include in a training step, so I am trying to achieve the same results by applying the same function sample-wise in the `get(...)` function in the `Dataset`

        adj = self._pad_adj(
            edge_index=edge_indices,
            edge_attr=edge_features,
            num_left_feat=left_features.shape[0],
            num_right_feat=right_features.shape[0],
            fill=0,
        )

        graph = BipartiteNodeData(
            edge_index=adj.indices(),
            edge_attr=adj.values(),
        )
However, results are not the same,

Here are the `indices()` and `edge_index` from both implementations 
```
>>> adj.indices()
tensor([[    0,     0,     0,  ..., 31247, 31247, 31247],
        [15516, 15517, 15592,  ..., 14718, 15307, 15417]])
>>> edge_index
tensor([[    0,     0,     0,  ..., 31247, 31247, 31247],
        [ 2118,  2119,  2194,  ..., 28883, 29472, 29582]])
```

Although it may seem that `edge_index[0]` values are equal to those of `adj.indices()[0]`, they're actually not. 

Note:

`_pad_adj` basically adds zeros to the diagonal and puts the features on the right top quadrant of the matrix and its transpose to the left bottom quadrant.

Any thoughts? Thanks.

**Comment by rusty1s** [comment]
Hey, I answered you on Slack as well, but just for completeness:

Batch-wise adjacency matrices are stacked diagonally, so converting your batched adjacency matrix to a dense one and then applying modifications to `adj[:num_vars, num_vars:]` will not work as intended.

I am also wondering why you do
```
adj = fill * np.ones((num_nodes, num_nodes))
```
since this will result in a sparse matrix with 0% sparsity.
